[KBFS-1871] Base journal disk limits on available disk space

akalin-keybase opened the pull request:
Add code to determine available disk space for a given path.

Use that to set the journal disk limit to
min(25% of available disk space, 50 GiB).

Make EnableJournaling return an error.

Use pkg/errors and %+v in more places.

akalin-keybase commented on libkbfs/config_local.go:
```diff
@@ -912,33 +911,44 @@ func (c *ConfigLocal) EnableJournaling(
+		return err
+	}
+	// Set journalDiskLimit to a quarter of the free disk space,
+	// up to 50 GiB.
```
I know you suggested 100 GB, but let's start with 50 for now.

akalin-keybase commented on libkbfs/disk_limits_unix.go:
```diff
@@ -0,0 +1,31 @@
+// TODO: Also return available files.
+func getDiskLimits(path string) (availableBytes uint64, err error) {
+	var stat unix.Statfs_t
+	err = unix.Statfs(path, &stat)
```
implemented this ourselves, since the package you suggested had somewhat outdated (used the deprecated syscall package) and sloppy (no error checking) code.

akalin-keybase commented on libkbfs/disk_limits_windows.go:
```diff
@@ -0,0 +1,38 @@
+		return 0, errors.WithStack(err)
+	}
+
+	dll := windows.NewLazySystemDLL("kernel32.dll")
```
According to the package docs, the method used by go-disk-usage is vulnerable to DLL preloading attacks.

akalin-keybase commented on libkbfs/kbfs_cr_test.go:
```diff
@@ -24,12 +24,12 @@ func readAndCompareData(t *testing.T, config Config, ctx context.Context,
 	fileNode, _, err := kbfsOps.Lookup(ctx, rootNode, "a")
 	if err != nil {
-		t.Fatalf("Couldn't lookup file: %v", err)
+		t.Fatalf("Couldn't lookup file: %+v", err)
```
The only change in this file is '%v", err' -> '%+v", err', and handling the errors from `EnableJournaling`.

akalin-keybase commented:
Oops, looks like I need to make sure the path passed to `getDiskLimits` is an existing one. I'll do that, but I think this PR is still reviewable.

akalin-keybase commented:
Okay, added a MkdirAll call.

akalin-keybase commented:
CI now looks like an unrelated chat failure.

strib commented on libkbfs/disk_limits_unix.go:
```diff
@@ -0,0 +1,31 @@
+
+	// TODO: Use stat.Ffree to return availableFiles.
+
+	return availableBytes, nil
```
I'm conflicted about whether the disk limit should be based on available bytes, rather than total bytes.  The argument against using available bytes is that it will change each time we restart KBFS, and the available bytes will be reduced by anything that's already in the journal on restart!  So I find it hard to reason about in the long run.  Of course using total bytes has the issue that the disk could already be full, but I'm inclined to say there's not much we can do about that, since the disk could fill up anyway after we start and we have to handle it with a nice error somehow anyway.  What do you think?

akalin-keybase replied:
I vote for available bytes, since using total bytes can lead to bad behavior with small disks, even when the disk isn't full. For example, a 10 GB disk that's almost full, say, only 2 GB free, would have all that remaining space gobbled up by the journal with a 25% limit.

I didn't think of the fact that available bytes already counts the journal space, but fortunately that's easy enough to fix -- if we want J <= p(J + A), then that's equivalent to requiring J <= (p/(1-p)) A; in this case, if we want p = 25%, then it suffices to require that the initial disk limit be 33 1/3% of the current available disk space. (This took me embarrassingly long to work out.) This is all approximate since we're only counting blocks, but we'll have to handle disk full errors separately.

What do you think?

akalin-keybase replied:
Ugh, scratch that last comment -- my analysis is bogus. In "J <= p(J + A)", the left and right Js aren't the same! The left one is the journal size over time, and the right one is the initial size, so more like "J(t) <= p(J(0) + A)"

akalin-keybase replied:
Okay, so I fixed this PR to take into account journal size when doing the disk limit calculation.

strib replied:
It's a tough call.  The problem with your argument is that any application on the computer can gobble up the available disk space, leaving us with just 1 KB that we instantly eat up.  I feel like the same argument cuts both ways, but by using the total space at least we maintain a consistent size that doesn't change on every restart.

I'll look through the code again now and think about it some more.  I might also ask Coyne for his opinion.

strib commented on libkbfs/disk_limits_unix.go:
```diff
@@ -0,0 +1,31 @@
+// the given path.
+//
+// TODO: Also return available files.
+func getDiskLimits(path string) (availableBytes uint64, err error) {
```
Any reason not to put this in `util.go`?  I slightly prefer to have fewer files, but don't care too much.

akalin-keybase replied:
Just the fact that it has platform-specific bits, and the fact that it's morally its own package (and may be in the future).

strib commented on libkbfs/disk_limits_windows.go:
```diff
@@ -0,0 +1,38 @@
+	proc := dll.NewProc("GetDiskFreeSpaceExW")
+	r1, _, err := proc.Call(uintptr(unsafe.Pointer(pathPtr)),
+		uintptr(unsafe.Pointer(&availableBytes)), 0, 0)
+	if r1 == 0 {
```
Not checking `err` here explicitly seems like it will make some future linter upset.

akalin-keybase replied:
I added a comment, and explicitly nil out the err, which should hopefully quiet the future linter.

strib commented on libkbfs/init.go:
```diff
@@ -486,8 +486,11 @@ func Init(ctx Context, params InitParams, keybaseServiceCn KeybaseServiceCn, onI
+		err := config.EnableJournaling(params.WriteJournalRoot,
 			params.TLFJournalBackgroundWorkStatus)
+		if err != nil {
+			return nil, err
```
Did you fix the race preventing journaling from coming up sometimes?  I'm worried that if the race is still there, this will break KBFS for existing users in hard-to-discover ways, even before journaling is on by default.

akalin-keybase replied:
This is now rebased on the PR that fixes the race.

akalin-keybase commented on libkbfs/config_local.go:
```diff
@@ -912,33 +912,50 @@ func (c *ConfigLocal) EnableJournaling(
+	// Set journalDiskLimit to a quarter of the free disk space,
+	// up to 50 GiB.
+	const maxJournalDiskLimit uint64 = 50 * 1024 * 1024 * 1024
+	var journalDiskLimit = availableBytes / 4
```
Fixed.

akalin-keybase commented:
Okay, this is ready for another look.

akalin-keybase commented:
Hold off on reviewing this until I fix the available size calculation.

akalin-keybase commented:
Okay, I still want to write some tests, but this is ready for another review. Most of the meat is in `semaphore_disk_limiter.go`.

strib commented on libkbfs/semaphore_disk_limiter.go:
```diff
@@ -0,0 +1,110 @@
+	if newByteLimit > oldByteLimit {
+		s.s.Release(newByteLimit - oldByteLimit)
+	} else if newByteLimit < oldByteLimit {
+		s.s.ForceAcquire(oldByteLimit - newByteLimit)
```
I feel like it's bad to let this force the semaphore into negative territory, at least while `s.availableBytes > 0` (or some other small limit).  We don't want it to suddenly start failing writes just because some other application dumped a bunch of data onto disk, as long as there's some disk space left.

I think the best way to do this depends on whatever is coming in your backpressure PR.  What we want is a smoothish backpressure function, instead of a hard stop whenever we update `availableBytes`.  Maybe for now we can do something like this?

```go
n := s.s.ForceAcquire(oldByteLimit - newByteLimit)
if n < 0 {
    // If we had room before, make sure we still have some minimum room now.
    // But don't reclaim past what the old limit was.
    reclaim := 10*1024*1024 - n  // Try to keep at least 10MB of available space.
    if reclaim > (oldByteLimit - newByteLimit) {
        reclaim = oldByteLimit - newByteLimit
    }
    s.s.Release(reclaim)
    s.byteLimit += reclaim
}
```

(Maybe it's better to build this into `kbfssync.Semaphore` so it's not racy, since this way some writes could fail before the `Release()` call happens.  Or you could grab the lock for reading during `beforeBlockPut`.)

strib commented on libkbfs/semaphore_disk_limiter.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	oldByteLimit := s.byteLimit
+	s.byteLimit = newByteLimit
```
We probably want to log the new limit whenever it changes (and thus have an `s.log` that's passed in on construction).

strib commented on libkbfs/semaphore_disk_limiter_test.go:
```diff
@@ -0,0 +1,30 @@
+
+	s.onUpdateAvailableBytes(0)
+	require.Equal(t, int64(0), s.getByteLimit())
+	s.onJournalEnable(50)
```
Can you add some tests for what happens when you call `onUpdateAvailableBytes()` after `onJournalEnable()`?

strib commented on libkbfs/tlf_journal.go:
```diff
@@ -366,10 +366,10 @@ func makeTLFJournal(
-			"Force-acquired %d bytes for %s: available=%d",
-			storedBytes, tlfID, availableBytes)
+			"Force-acquired %d bytes for %s: %v",
+			storedBytes, tlfID, j.diskLimiter)
```
Why not continue to print out `availableBytes`?  Was this for debugging? Doesn't seem too useful to print out the interface here, since there's no `String()` on the `diskLimiter` interface.

akalin-keybase commented:
I'm going to hold off on this PR until #722 is merged, since it seems that it makes more sense to get that in first anyway.

akalin-keybase commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+	// We don't actually look at blockBytes -- we're assuming that
+	// it's small compared to the other numbers.
+
+	availBytes, err := s.availBytesFn()
```
Do you think this is too expensive to call on every block put? If so, I can do something like only update this every n seconds, and in between approximate it by subtracting put blocks in afterBlockPut.

akalin-keybase commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
 	}
 
-	return s.s.Acquire(ctx, blockBytes)
+	return 0, nil
```
We probably don't even really need a semaphore here. If the journal reaches the limits, it'll just be super slow. What do you think?

strib replied:
That would make it possible to exceed the limit though, right?  Shouldn't we make sure somewhere we don't exceed it, even after a delay?  If there isn't any flushing happening, say because it's offline, we don't want to fill the disk.

akalin-keybase commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+		return s.journalBytes
+	}()
+
+	r := math.Max(
```
Here's the important part

akalin-keybase commented:
I've effectively rewritten this, so this needs a fresh look. All the core logic is there, the WIP is just referring to adding comments, tests, etc.

strib commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
-	if usedBytes >= s.backpressureMaxThreshold {
-		return s.maxDelay
-	}
+func boundedLerp(x, min, max float64) float64 {
```
Unused?

akalin-keybase replied:
Removed.

strib commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+func (s *backpressureDiskLimiter) onBlockDelete(blockBytes int64) {
+	s.journalBytesLock.Lock()
+	defer s.journalBytesLock.Unlock()
+	s.journalBytes -= blockBytes
```
Do we want this to interrupt any delay functions to recalculate?  No sense in waiting a few 11 second delay if the journal size shrinks massively after 50ms.

akalin-keybase replied:
I thought about it, and I don't think it'll help outside of some edge cases (e.g., disk is almost full, user deletes a bunch of stuff). Added a TODO, though.

strib commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+	m := s.backpressureMinThreshold
+	M := s.backpressureMaxThreshold
+	scale := math.Min(1.0, math.Max(0.0, (r-m)/(M-m)))
+	return time.Duration(scale * float64(s.maxDelay))
```
If we wait a full `maxDelay` on every block of a multi-block put, we'll run into OS timeouts pretty quickly.  However, since blocks are mostly put in parallel, outside of the journal lock, I guess in practice we're actually waiting only one `maxDelay` as long as there aren't more than 100 blocks put at once.

Still, maybe it's safer to pass in `ctx` here and scale  `min(s.maxDelay, ctx.Deadline().Sub(time.Now())-1*time.Second)` or something similar.

akalin-keybase replied:
Good point, done.

strib commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
 	if blockBytes == 0 {
 		// Better to return an error than to panic in Acquire.
-		return s.s.Count(), errors.New(
+		return 0, errors.New(
```
Did you mean to not use `s.s.Count()` throughout this function?

akalin-keybase replied:
Yeah, since I removed the semaphore. But now that it's back, added it back in.

akalin-keybase commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+
+	r := math.Max(
+		float64(journalBytes)/float64(s.maxJournalBytes),
+		float64(journalBytes)/(float64(journalBytes)+float64(availBytes)))
```
I wanted to cast it first to avoid overflow. Pulled it out into some vars and added a comment.

akalin-keybase commented:
I can't seem to reply to some of your other comments, so doing it here.

I added a benchmark for `getDiskLimits()` and am getting ~600ns/call (mac, linux), so it looks like it's not worth optimizing.

Added the semaphore back in.

This is ready for another high-level pass through the disk limiter (not a full pass yet).

strib commented:
Disk limiter high-level pass looks good to me.

akalin-keybase commented on libkbfs/backpressure_disk_limiter_test.go:
```diff
@@ -5,36 +5,344 @@
+
+// TestBackpressureDiskLimiterLargeDiskDelay checks the delays when
+// pretending to have a small disk.
+func TestBackpressureDiskLimiterSmallDisk(t *testing.T) {
```
I need to flesh this test out a bit more.

akalin-keybase replied:
I'll do this in the followup.

akalin-keybase commented on libkbfs/backpressure_disk_limiter.go:
```diff
@@ -14,42 +16,81 @@ import (
+// Let J be the (approximate) byte usage of the journal and F be the
+// free bytes on disk. Then we want to enforce
+//
+//   J <= min(J+F, L),
```
One thing I forgot to do was to limit to a *fraction* of the available disk space (i.e., 25%) -- here we're limiting to 100% of the available disk space. But I think this is still reviewable.

strib replied:
Ah yeah, we can do that in a followup, but should probably be in before we go live internally.

akalin-keybase replied:
Yeap, agree.

akalin-keybase commented:
Okay, this is ready for a full review.

strib commented on libkbfs/backpressure_disk_limiter_test.go:
```diff
@@ -5,36 +5,344 @@
+	checkCounters(0)
+}
+
+// TestBackpressureDiskLimiterLargeDiskDelay checks the delays when
```
Wrong comment.

akalin-keybase replied:
Fixed.
